File: iOS/Objective C/LevelPlayDemoTests/RewardedIntegrationTest.m
#import <XCTest/XCTest.h>
#import "DemoViewController.h"

@interface RewardedIntegrationTest : XCTestCase <LPMRewardedAdDelegate, LPMImpressionDataDelegate>
@property (nonatomic) XCTestExpectation *sdkInitExpectation;
@property (nonatomic) XCTestExpectation *loadExpectation;
@property (nonatomic) XCTestExpectation *impressionExpectation;
@property (nonatomic) LPMRewardedAd *rewardedAd;
@end

@implementation RewardedIntegrationTest

- (void)testInitLoadAndImpression {
    self.sdkInitExpectation = [self expectationWithDescription:@"Init"];
    self.loadExpectation = [self expectationWithDescription:@"Load"];
    self.impressionExpectation = [self expectationWithDescription:@"Impression"];

    [LevelPlay addImpressionDataDelegate:self];

    LPMInitRequestBuilder *builder = [[LPMInitRequestBuilder alloc] initWithAppKey:kAppKey];
    LPMInitRequest *request = [builder build];

    [LevelPlay initWithRequest:request completion:^(LPMConfiguration *config, NSError *error) {
        XCTAssertNil(error, @"Init failed: %@", error);
        [self.sdkInitExpectation fulfill];

        self.rewardedAd = [[LPMRewardedAd alloc] initWithAdUnitId:kRewardedAdUnit];
        [self.rewardedAd setDelegate:self];
        [self.rewardedAd loadAd];
    }];

    [self waitForExpectations:@[self.sdkInitExpectation] timeout:10];
    [self waitForExpectations:@[self.loadExpectation] timeout:15];

    UIViewController *rootVC = UIApplication.sharedApplication.keyWindow.rootViewController;
    [self.rewardedAd showAdWithViewController:rootVC placementName:nil];

    [self waitForExpectations:@[self.impressionExpectation] timeout:20];
}

#pragma mark - LPMRewardedAdDelegate

- (void)didLoadAdWithAdInfo:(LPMAdInfo *)adInfo {
    [self.loadExpectation fulfill];
}

- (void)didFailToLoadAdWithAdUnitId:(NSString *)adUnitId error:(NSError *)error {
    XCTFail(@"Load failed: %@", error);
}

- (void)didDisplayAdWithAdInfo:(LPMAdInfo *)adInfo {}
- (void)didFailToDisplayAdWithAdInfo:(LPMAdInfo *)adInfo error:(NSError *)error {}
- (void)didClickAdWithAdInfo:(LPMAdInfo *)adInfo {}
- (void)didCloseAdWithAdInfo:(LPMAdInfo *)adInfo {}
- (void)didRewardAdWithAdInfo:(LPMAdInfo *)adInfo reward:(LPMReward *)reward {}
- (void)didChangeAdInfo:(LPMAdInfo *)adInfo {}

#pragma mark - LPMImpressionDataDelegate

- (void)impressionDataDidSucceed:(LPMImpressionData *)impressionData {
    [self.impressionExpectation fulfill];
}

@end

File: iOS/Objective C/LevelPlayDemoTests/RewardedIntegrationTest.swift
import Testing
import IronSource
import UIKit

@Test(.timeLimit(.minutes(1)))
@MainActor
func testRewardedInitLoadAndImpression() async throws {
    let delegate = RewardedTestDelegate()
    let impressionDelegate = RewardedImpressionDelegate()

    LevelPlay.add(impressionDelegate as LPMImpressionDataDelegate)

    // Init SDK
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
        let requestBuilder = LPMInitRequestBuilder(appKey: kAppKey)
        let initRequest = requestBuilder.build()
        LevelPlay.initWith(initRequest) { config, error in
            if let error = error {
                continuation.resume(throwing: error)
            } else {
                continuation.resume()
            }
        }
    }

    // Create and load ad
    let rewardedAd = LPMRewardedAd(adUnitId: kRewardedAdUnit)
    rewardedAd.setDelegate(delegate)

    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
        delegate.onLoad = { continuation.resume() }
        delegate.onLoadFail = { error in continuation.resume(throwing: error) }
        rewardedAd.loadAd()
    }

    // Show ad
    let rootVC = UIApplication.shared.keyWindow?.rootViewController
    rewardedAd.showAd(viewController: rootVC!, placementName: nil as String?)

    // Wait for impression
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
        impressionDelegate.onImpression = { continuation.resume() }
    }
}

private class RewardedTestDelegate: NSObject, LPMRewardedAdDelegate {
    var onLoad: (() -> Void)?
    var onLoadFail: ((Error) -> Void)?

    func didLoadAd(with adInfo: LPMAdInfo) { onLoad?() }
    func didFailToLoadAd(withAdUnitId adUnitId: String, error: Error) { onLoadFail?(error) }
    func didDisplayAd(with adInfo: LPMAdInfo) {}
    func didFailToDisplayAd(with adInfo: LPMAdInfo, error: Error) {}
    func didClickAd(with adInfo: LPMAdInfo) {}
    func didCloseAd(with adInfo: LPMAdInfo) {}
    func didRewardAd(with adInfo: LPMAdInfo, reward: LPMReward) {}
    func didChangeAdInfo(_ adInfo: LPMAdInfo) {}
}

private class RewardedImpressionDelegate: NSObject, LPMImpressionDataDelegate {
    var onImpression: (() -> Void)?

    func impressionDataDidSucceed(_ impressionData: LPMImpressionData) {
        onImpression?()
    }
}
